design-proposals/shadow-node.md

```diff
@@ -0,0 +1,328 @@
+    - the storage kubevirt takes.
+- Adding virt-controller as a "proxy" agent, we now depend on its availability. This could cause an increase in heartbeat timeout in corner cases like upgrade.
+- To a degree, it is still possible to manipulate pods by causing VMs to migrate from other nodes.
+
+### API Examples
+```yaml
+apiVersion: <apiversion>
+kind: ShadowNode
+metadata:
+  name: <Equivalent to existing name>
+  # Labels and Annotations only
+spec: {} //Empty, allows further extension
+status: {} //Empty, allows further extension
+```
+
+### Solution option #3: Creating a REST/GRPC communication between virt-handler and virt-controller
+Pass the labels/annotations through a REST/GRPC channel to the virt-controller. REST option is preferred because the server-side is already implemented.
+
+Virt-handler (Server): Instead of directly patching labels/annotation to the node, Virt-handler will maintain a "cache" (NodeInfoMgr) where all the components will "set" their data.
+Virt handler can use its existing [web-service](https://github.com/kubevirt/kubevirt/blob/b102c56f0fcd52feff3ff7a6296737b8e8b99131/cmd/virt-handler/virt-handler.go#L556) to expose one more endpoint - GetNodeInfo. Once reached - this endpoint will return the node data.
+
+Virt-Controller (Client): Add a new web-service client infra that will:
+1. run as a go routine and poll the client for updates every one minute.This frequency is chosen because it is the highest of the frequencies (=heartbeat).
+2. ask every Virt-handler server separately for the NodeInfo (need to make sure it's possible to address them in a unique manner).
+3. filter non-kubevirt owned annotations/labels.
+4. patch the node if necessary.
+
+Notes:
+- The heartbeat annotation and virt-controller's [Node-controller](https://github.com/kubevirt/kubevirt/blob/b102c56f0fcd52feff3ff7a6296737b8e8b99131/pkg/virt-controller/watch/node.go) can be removed in favor using the web-service timeout mechanism (if received timeout - set node as unreachable).
+
+Pros:
+- Avoid introducing a new CRD to the project.
+- Using existing web-server infra on the server side.
+- security:
+  - the web-service is secure using the cert-manager internal API.
+  - Virt-handler does not need to access kube-API at all, and cannot modify unrelated nodes or shadow nodes. Virt-controller does all the risky work
+- Scalability:
+  - Reduces #API requests (i.e. node patches) to a minimum.
+
+Cons:
+- More complex. It's easier and more appealing to use the k8s API, i.e. patching labels/annotations, than to implement and maintain GRPC/REST case logic internally.
+- Virt-handler will no longer be able to immediately set the node as unschedulable (in case of virt-handler SIGTERM). However, this use-case is pretty rare, and IMO should be reconsidered.
```

Comment:
Please correct the second sentence.

Reply:
Added more info on this both on the Con and on the solution body.